### PR TITLE
Create custom DiagnosticsNode type for DevTools deep links.

### PR DESCRIPTION
**Pull request by kenzieschmoll**
IDEs will look for this specific diagnostic node type in order to prompt the user to inspect the errors in DevTools.
cc @helin24 @DanTup

### CODE REVIEW

File: packages/flutter/lib/src/foundation/diagnostics.dart [hunk 1]
```diff
@@ -1801,6 +1801,19 @@ class MessageProperty extends DiagnosticsProperty<void> {
...
 }
 
+/// Debugging message for DevTools deep links.
+class DevToolsDeepLinkDiagnosticsNode extends StringProperty {
+  /// Create a diagnostics property that displays a deep link to DevTools.
+  ///
+  /// [value] will return the String representation of the DevTools url.
+  ///
+  /// The [description] and [value] arguments must not be null.
+  DevToolsDeepLinkDiagnosticsNode(String description, String value) :
```
Comment (jacob314) [inline packages/flutter/lib/src/foundation/diagnostics.dart:1811]: Consider renaming parameter `value` to `url` and making it named.

File: packages/flutter/lib/src/foundation/diagnostics.dart [hunk 2]
```diff
@@ -1801,6 +1801,19 @@ class MessageProperty extends DiagnosticsProperty<void> {
        super(name, null, description: message, style: style, level: level);
 }
 
+/// Debugging message for DevTools deep links.
```
Comment (jacob314) [inline packages/flutter/lib/src/foundation/diagnostics.dart:1804]: Top level comments should ideally say something more than just the words from the class name in another order. Clarify why it makes sense to indicate that a link is to Flutter DevTools vs any other deep link you might have in the app.
Change the class name to be consistent with the other class names.
For example
DevToolsDeepLinkProperty would be better.

Comment (jacob314) [inline packages/flutter/lib/src/foundation/diagnostics.dart:1804]: Move this class from diagnostics.dart to widget_inspector.dart. This isn't general purpose enough to be in foundation

File: packages/flutter/lib/src/widgets/widget_inspector.dart [hunk 3]
```diff
@@ -2934,8 +2934,9 @@ Iterable<DiagnosticsNode> _describeRelevantUserCode(Element element) {
       final String? devToolsInspectorUri =
           WidgetInspectorService.instance._devToolsInspectorUriForElement(target);
       if (devToolsInspectorUri != null) {
-        devToolsDiagnostic = DiagnosticsNode.message(
+        devToolsDiagnostic = DevToolsDeepLinkDiagnosticsNode(
```
Comment (jacob314) [inline packages/flutter/lib/src/widgets/widget_inspector.dart:2937]: consider moving some of the logic from _devToolsInspectorUriForElement into DevToolsDeepLink as it is tightly coupled and that would make the DevToolsDeepLink class a little less surprising than if its public constructor takes an arbitrary URL which might or might not be from DevTools.

File: packages/flutter/lib/src/foundation/diagnostics.dart [hunk 4]
```diff
@@ -1801,6 +1801,19 @@ class MessageProperty extends DiagnosticsProperty<void> {
...
+class DevToolsDeepLinkDiagnosticsNode extends StringProperty {
+  /// Create a diagnostics property that displays a deep link to DevTools.
+  ///
+  /// [value] will return the String representation of the DevTools url.
```
Comment (goderbauer) [inline packages/flutter/lib/src/foundation/diagnostics.dart:1808]: nit: doc sentences should be grammatically correct and start with a capital letter, e.g. "The [value] will..."

**Comment by kenzieschmoll** [comment]
Addressed review comments and refactored some things. Represented the diagnostic value as a map so I can add useful parameters for IDEs to consume (other wise they would have had to parse information from the url). PTAL.

File: packages/flutter/lib/src/widgets/widget_inspector.dart [hunk 5]
```diff
@@ -2958,6 +2940,115 @@ Iterable<DiagnosticsNode> _describeRelevantUserCode(Element element) {
...
+///
+/// This property encloses it [value] as a Map<String, String> object. This map
+/// is composed of the following information:
+///
+/// `url` is the String representation of the deep link to DevTools.
+/// `screenId` is an optional id specifying the Flutter DevTools screen that this
+/// diagnostic pertains to.
+/// `objectId` is an optional id specifying the data object in Flutter DevTools
+/// that this diagnostic pertains to.
+class DevToolsDeepLinkProperty extends DiagnosticsProperty<Map<String, String>> {
```
Comment (jacob314) [inline packages/flutter/lib/src/widgets/widget_inspector.dart:2954]: This should not be a property on Map<String, String>
instead make 
DevToolsDeepLink extend Diagnosticable.
Then it is clean enough to add the extra json you want and specify the tostring you needed.

File: packages/flutter/test/widgets/widget_inspector_test.dart [hunk 6]
```diff
@@ -2835,6 +2835,31 @@ class _TestWidgetInspectorService extends TestWidgetInspectorService {
...
+        equals(<String, dynamic>{
+          'description': 'description of the deep link',
+          'type': 'DevToolsDeepLinkProperty',
+          'name': '',
+          'style': 'singleLine',
+          'allowNameWrap': true,
+          'missingIfNull': false,
+          'propertyType': 'String',
+          'defaultLevel': 'info',
+          'value': 'the-deep-link',
```
Comment (jacob314) [inline packages/flutter/test/widgets/widget_inspector_test.dart:2859]: nit: make this
'http://the-deeplink/'

**Review by jacob314 — APPROVED** [review]
lgtm

File: packages/flutter/lib/src/widgets/widget_inspector.dart [hunk 7]
```diff
@@ -2958,6 +2959,27 @@ Iterable<DiagnosticsNode> _describeRelevantUserCode(Element element) {
   return nodes;
 }
 
+
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/widget_inspector.dart:2962]: nit: just one blank line

**Review by goderbauer — APPROVED** [review]
LGTM

**Pull request merged**